python/ray/dag/compiled_dag_node.py

```diff
@@ -405,24 +467,45 @@ def _get_or_compile(
                 frontier.append(idx)
 
         for node_idx, task in self.idx_to_task.items():
-            if node_idx == self.input_task_idx:
-                # We don't need to assign an actual task for the input node.
+            if node_idx in self.input_task_idxs:
+                # We don't need to assign an actual task for the input node(s).
                 continue
 
             if node_idx == self.output_task_idx:
-                # We don't need to assign an actual task for the input node.
+                # We don't need to assign an actual task for the output node.
                 continue
 
             resolved_args = []
+            # for single arg, read_idx is -1, meaning taking the whole input,
+            # for multiple args, read_idx means the index of element to take,
+            # Example:
+            # inp = InputNode()
+            # f.remote(inp.x)
+            # g.remote(inp.y)
+            # f and g both read (x, y) from the same channel, but pick different indices
+            read_idxs = []
             has_at_least_one_channel_input = False
             for arg in task.args:
                 if isinstance(arg, DAGNode):
-                    arg_idx = self.dag_node_to_idx[arg]
-                    arg_channel = self.idx_to_task[arg_idx].output_channel
+                    node_idx = self.dag_node_to_idx[arg]
+                    arg_channel = self.idx_to_task[node_idx].output_channel
                     assert arg_channel is not None
-                    resolved_args.append(arg_channel)
+                    if isinstance(arg, MultiInputNode):
+                        # Each MultiInputNode reference is responsible for
+                        # a single argument, corresponding to a non-negative
+                        # index in read_idx
+                        read_idx = len([i for i in read_idxs if i >= 0])
```

Comment:
I didn't understand this line -- shouldn't the read_idx be based on which InputAttributeNode was this task's arg?

Reply:
fixed!